### PR TITLE
feat: Tests for the simulator

### LINKED ISSUE #3161: Tests for the simulator
### Tests for the simulator: 
Simulator currently has a [testbench engine](https://github.com/CircuitVerse/CircuitVerse/blob/master/simulator/src/testbench.js) which serves as a good API for running tests on the simulator. The goal is to build a test suite on top of this.

### PR DESCRIPTION
**Pull request by pulkit-30**
## [GSOC'22 Task](https://github.com/CircuitVerse/CircuitVerse/wiki/GSoC%2722-Project-List#project-4---circuitverse-simulator-improvements)

Fixes #3161 

### Implemented Tasks

- [x]  Added tests for 
> - [x] AND gate
> - [x] NOR gate
> - [x] NOT gate
> - [x] XNOR gate
> - [x] XOR gate
> - [x] OR gate
> - [x] NAND gate

- [x]  Added workflow to [run tests](https://github.com/CircuitVerse/CircuitVerse/runs/6725951966?check_suite_focus=true) on Pull Request(only for the code changes inside simulator/ dir)

## Basics Steps to create a Test are:
1. Setup Simulator Environment using setup() function.
2. Load circuit Data using load() function.
3. Run test using testbench engine (runAll() function).

### All Tests Pass

https://user-images.githubusercontent.com/76155456/171418413-4e9a209a-6bc5-4933-a842-8281e0f65fee.mov

## Working
163724136-f7a6b647-2041-4837-a9ea-0ec8b3fda051

---

### CODE REVIEW

**Review by gr455 — APPROVED** [review]
Amazing! Looks great so far :tada: :100: 
PS. Super sorry for the late review. Busy with placements stuff

**Comment by pulkit-30** [comment]
> Amazing! Looks great so far 🎉 💯
PS. Super sorry for the late review. Busy with placements stuff

No problem, and Thanks for reviewing 🎉 
Can we initiate the merging process? becoz next PR will depend on these changes...

File: package.json [hunk 1]
```diff
@@ -47,6 +53,7 @@
     "esbuild-plugin-sass": "^1.0.1",
     "esbuild-rails": "^1.0.3",
     "file-loader": "^6.2.0",
+    "identity-obj-proxy": "^3.0.0",
```
Comment (tachyons) [inline package.json:56]: why do we need this on dependencies ?

File: package.json [hunk 2]
```diff
@@ -35,7 +39,9 @@
     "banana-i18n": "^2.3.3",
     "better-docs": "^2.7.2",
     "bootstrap": "^4.6.1",
+    "bootstrap-input-spinner": "^3.1.10",
     "bootstrap-tagsinput": "^0.7.1",
+    "canvas": "^2.9.1",
```
Comment (tachyons) [inline package.json:44]: this package seems to have additional dependencies like cairo, we may have to update documentation

File: package.json (hunk 1 above)
Comment (pulkit-30) [reply package.json:56]: [Errorr](https://gist.github.com/pulkit-30/9182f6875af1239515e1cc4a481b6f25#error---css-files)  for `importing CSS modules` -- `codemirror.css`

Error log:

```
    Details:

    /Users/pulkitgupta/Desktop/OpenSource/CircuitVerse/node_modules/codemirror/lib/codemirror.css:3
    .CodeMirror {
    ^

    SyntaxError: Unexpected token '.'

      3 |
      4 | import CodeMirror from 'codemirror/lib/codemirror.js';
    > 5 | import 'codemirror/lib/codemirror.css';
        | ^
      6 |
      7 | // Importing CodeMirror themes
      8 | import 'codemirror/theme/3024-day.css';

      at Runtime.createScriptFromCode (node_modules/jest-runtime/build/index.js:1773:14)
      at Object.<anonymous> (simulator/src/Verilog2CV.js:5:1)
```
[identity-obj-proxy](https://github.com/keyz/identity-obj-proxy#identity-obj-proxy---) solves this issue (simply say mock these modules while running test).

### Jest Config

```
.
.

    moduleNameMapper: {
        "\\.(css|less|sass|scss)$": "identity-obj-proxy",
        "typeface-nunito": "identity-obj-proxy",
    },
.
.

```


Open to an alternative solution

File: package.json (hunk 2 above)
Comment (pulkit-30) [reply package.json:44]: Sure, I will discuss this with the mentor on call 👍🏽

**Pull request merged**